### PR TITLE
Drop support for pre-8.0 GHCs

### PR DESCRIPTION
**Pull request by RyanGlScott**

### CODE REVIEW

File: src/Data/Functor/Classes/Compat.hs [hunk 1]
```diff
@@ -1,5 +1,3 @@
-#include "free-common.h"
-#ifdef LIFTED_FUNCTOR_CLASSES
 {-# LANGUAGE Safe #-}
 module Data.Functor.Classes.Compat (
```
Comment (RyanGlScott) [inline src/Data/Functor/Classes/Compat.hs:2]: I'm not sure how necessary this module is nowadays, given that it no longer defines any compatibility shims. Perhaps I will just fold this into `Control.Applicative.Free`.

File: src/Control/Monad/Trans/Free.hs [hunk 2]
```diff
@@ -212,80 +162,42 @@ free :: FreeF f a (Free f a) -> Free f a
...
-#endif
 
-#ifdef LIFTED_FUNCTOR_CLASSES
 instance (Read1 f, Read1 m, Read a) => Read (FreeT f m a) where
-#else
-instance (Functor f, Read1 f, Functor m, Read1 m, Read a) => Read (FreeT f m a) where
-#endif
   readsPrec = readsPrec1
 
 instance (Functor f, Monad m) => Functor (FreeT f m) where
```
Comment (RyanGlScott) [inline src/Control/Monad/Trans/Free.hs:203]: We could go even further and simplify the `Monad m` constraint to `Functor m`, but doing so would require some code changes. I'll defer that to a future PR.

File: free.cabal [hunk 3]
```diff
@@ -82,46 +73,22 @@ library
...
-  if impl(ghc >=7.8)
-    build-depends:
-      profunctors >= 5.6.1 && < 6
-  else
-    build-depends:
-      profunctors >= 5.2.2 && < 5.3
+    template-haskell     >= 2.11    && < 2.20
 
   if !impl(ghc >= 8.2)
-    build-depends: bifunctors >= 5.5.9 && < 6
```
Comment (turion) [inline free.cabal:115]: That's a change. Now it doesn't depend on `bifunctors` at all anymore?

File: src/Control/Alternative/Free.hs [hunk 4]
```diff
@@ -47,14 +39,8 @@ infixl 3 `Ap`
 data AltF f a where
   Ap     :: f a -> Alt f (a -> b) -> AltF f b
   Pure   :: a                     -> AltF f a
-#if __GLASGOW_HASKELL__ >= 707
-  deriving Typeable
```
Comment (turion) [inline src/Control/Alternative/Free.hs:51]: Are you sure you want to remove the `Typeable` instance? The trivial change would be to keep it unconditionally.

Comment (turion) [reply src/Control/Alternative/Free.hs:51]: (Similarly in all other analogous places)

File: src/Control/Monad/Trans/Free/Ap.hs [hunk 5]
```diff
@@ -60,31 +54,19 @@ import Control.Monad.State.Class
...
-#if !(MIN_VERSION_base(4,8,0))
-import Data.Foldable
-import Data.Monoid
-#endif
 
 -- | The base functor for a free monad.
 data FreeF f a b = Pure a | Free (f b)
-  deriving (Eq,Ord,Show,Read
-#if __GLASGOW_HASKELL__ >= 707
-           ,Typeable ,Generic, Generic1
```
Comment (turion) [inline src/Control/Monad/Trans/Free/Ap.hs:83]: Similarly, I believe you'd want to derive `Typeable` here.

File: free.cabal (hunk 3 above)
Comment (RyanGlScott) [reply free.cabal:115]: This was replaced with the `bifunctor-classes-compat` library:

https://github.com/ekmett/free/blob/8f8b5e8734ef3ba97278789debed4962c1ea8279/free.cabal#L91

This was recently split out of `bifunctors` to provide all of the compatibility shims for `Bifunctor`, `Bifoldable`, and `Bitraversable` without any of the other machinery in `bifunctors` (which incurs additional dependencies). The next `bifunctors` release will also depend on `bifunctor-classes-compat`. See ekmett/semigroupoids#130 for the discussion that originally motivated this.

Strictly speaking, I don't have to replace `bifunctors` with `bifunctor-classes-compat` in this particular spot, but I figured that we might as well while we're in the neighborhood.

File: src/Control/Alternative/Free.hs (hunk 4 above)
Comment (RyanGlScott) [reply src/Control/Alternative/Free.hs:51]: I actually didn't remove it, as GHC automatically defines `Typeable` instances for all type and data constructors from 7.10 onwards. In fact, the `Data` instances defined elsewhere in this library wouldn't typecheck without the auto-defined `Typeable` instances!

Comment (turion) [reply src/Control/Alternative/Free.hs:51]: Ah, I wasn't aware, thanks!

**Review by turion — APPROVED** [review]

**Pull request merged**